### ISSUE TITLE
Incremental compilation is completely broken in 1.3 on Linux

### ISSUE THREAD

**Issue by smarter** [sbt/sbt#4543]
To reproduce this, I've created the simplest possible project and set build.properties to the latest sbt nightly: https://github.com/smarter/sbt-inc-broken

```scala
% sbt

[info] Loading project definition from /home/smarter/opt/sbt-inc-broken/project
[info] Loading settings for project root from build.sbt ...
[info] Set current project to root (in build file:/home/smarter/opt/sbt-inc-broken/)
[info] sbt server started at local:///home/smarter/.sbt/1.0/server/c00128c57db581cfabf9/sock
[root@master]> compile
[info] Compiling 1 Scala source to /home/smarter/opt/sbt-inc-broken/target/scala-2.12/classes ...
[info] Done compiling.
[success] Total time: 1 s, completed 12 févr. 2019 23:13:54
[root@master]> compile
[info] Compiling 1 Scala source to /home/smarter/opt/sbt-inc-broken/target/scala-2.12/classes ...
[info] Done compiling.
[success] Total time: 0 s, completed 12 févr. 2019 23:13:56
[root@master]> compile
[info] Compiling 1 Scala source to /home/smarter/opt/sbt-inc-broken/target/scala-2.12/classes ...
[info] Done compiling.
[success] Total time: 0 s, completed 12 févr. 2019 23:13:57
```

Turning on debug logs doesn't help much:
```scala
[root@master]> debug
[root@master]> set incOptions in Global ~= { _.withApiDebug(true).withRelationsDebug(true) }
[root@master]> compile
[debug] Updating ...
[... A ton of useless overly verbose update messages ...]
[debug] Done updating 
[debug] 
[debug] Initial source changes: 
[debug]         removed:Set()
[debug]         added: Set()
[debug]         modified: Set()
[debug] Invalidated products: Set(/home/smarter/opt/sbt-inc-broken/target/scala-2.12/classes/A.class)
[debug] External API changes: API Changes: Set()
[debug] Modified binary dependencies: Set()
[debug] Initial directly invalidated classes: Set()
[debug] 
[debug] Sources indirectly invalidated by:
[debug]         product: Set(/home/smarter/opt/sbt-inc-broken/A.scala)
[debug]         binary dep: Set()
[debug]         external source: Set()
[debug] All sources are invalidated.
[debug] [inv] Invalidate package objects by inheritance only...
[debug] Initial set of included nodes: 
[debug] [inv] Package object invalidations: 
[debug] Recompiling all sources: number of invalidated sources > 50.0% of all sources
[debug] [inv] ********* Pruned: 
[debug] [inv] Relations (with name hashing enabled):
[debug] [inv]   products: Relation [ ]
[debug] [inv]   library deps: Relation [ ]
[debug] [inv]   library class names: Relation [
[debug] [inv]     /usr/lib/jvm/java-8-openjdk-amd64/jre/lib/rt.jar -> java.lang.Object
[debug] [inv] ]
[debug] [inv]   internalDependencies: 
[debug] [inv]     
[debug] [inv]   externalDependencies: 
[debug] [inv]     
[debug] [inv]   class names: Relation [ ]
[debug] [inv]   used names: Relation [ ]
[debug] [inv]   product class names: Relation [ ]
[debug] [inv] *********
[info] Compiling 1 Scala source to /home/smarter/opt/sbt-inc-broken/target/scala-2.12/classes ...
[debug] Getting org.scala-sbt:compiler-bridge_2.12:1.2.1-bin-20190204T212322:compile for Scala 2.12.8
[debug] Getting org.scala-sbt:compiler-bridge_2.12:1.2.1-bin-20190204T212322:compile for Scala 2.12.8
[debug] [zinc] Running cached compiler 3d250bda for Scala compiler version 2.12.8
[debug] [zinc] The Scala compiler is invoked with:
[debug]         -bootclasspath
[debug]         /usr/lib/jvm/java-8-openjdk-amd64/jre/lib/resources.jar:/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/rt.jar:/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/sunrsasign.jar:/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/jsse.jar:/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/jce.jar:/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/charsets.jar:/usr/lib/jvm/java-8-openjdk-amd64/jre/lib/jfr.jar:/usr/lib/jvm/java-8-openjdk-amd64/jre/classes:/home/smarter/.sbt/boot/scala-2.12.8/lib/scala-library.jar
[debug]         -classpath
[debug]         /home/smarter/opt/sbt-inc-broken/target/scala-2.12/classes
[debug] Scala compilation took 0.122393405 s
[info] Done compiling.
[debug] [inv] ********* Fresh: 
[debug] [inv] Relations (with name hashing enabled):
[debug] [inv]   products: Relation [
[debug] [inv]     A.scala -> target/scala-2.12/classes/A.class
[debug] [inv] ]
[debug] [inv]   library deps: Relation [
[debug] [inv]     A.scala -> /usr/lib/jvm/java-8-openjdk-amd64/jre/lib/rt.jar
[debug] [inv] ]
[debug] [inv]   library class names: Relation [
[debug] [inv]     /usr/lib/jvm/java-8-openjdk-amd64/jre/lib/rt.jar -> java.lang.Object
[debug] [inv] ]
[debug] [inv]   internalDependencies: 
[debug] [inv]     
[debug] [inv]   externalDependencies: 
[debug] [inv]     
[debug] [inv]   class names: Relation [
[debug] [inv]     A.scala -> A
[debug] [inv] ]
[debug] [inv]   used names: Relation [
[debug] [inv]     A -> UsedName(A,[Default])
[debug] [inv]     A -> UsedName(AnyRef,[Default])
[debug] [inv]     A -> UsedName(Object,[Default])
[debug] [inv]     A -> UsedName(java;lang;Object;init;,[Default])
[debug] [inv]     A -> UsedName(scala,[Default])
[debug] [inv] ]
[debug] [inv]   product class names: Relation [
[debug] [inv]     A -> A
[debug] [inv] ]
[debug] [inv] *********
[debug] [inv] ********* Merged: 
[debug] [inv] Relations (with name hashing enabled):
[debug] [inv]   products: Relation [
[debug] [inv]     A.scala -> target/scala-2.12/classes/A.class
[debug] [inv] ]
[debug] [inv]   library deps: Relation [
[debug] [inv]     A.scala -> /usr/lib/jvm/java-8-openjdk-amd64/jre/lib/rt.jar
[debug] [inv] ]
[debug] [inv]   library class names: Relation [
[debug] [inv]     /usr/lib/jvm/java-8-openjdk-amd64/jre/lib/rt.jar -> java.lang.Object
[debug] [inv] ]
[debug] [inv]   internalDependencies: 
[debug] [inv]     
[debug] [inv]   externalDependencies: 
[debug] [inv]     
[debug] [inv]   class names: Relation [
[debug] [inv]     A.scala -> A
[debug] [inv] ]
[debug] [inv]   used names: Relation [
[debug] [inv]     A -> UsedName(A,[Default])
[debug] [inv]     A -> UsedName(AnyRef,[Default])
[debug] [inv]     A -> UsedName(Object,[Default])
[debug] [inv]     A -> UsedName(java;lang;Object;init;,[Default])
[debug] [inv]     A -> UsedName(scala,[Default])
[debug] [inv] ]
[debug] [inv]   product class names: Relation [
[debug] [inv]     A -> A
[debug] [inv] ]
[debug] [inv] *********
```

However, if I restart sbt, then `compile` succeeds without doing anything, right until I do `clean` once, then it's back to `compile` compiling everything all the time.

I've tried manually bisecting this by using different sbt nightly builds, but downloading  the jars for a new nightly takes several minutes (!) and the couple I tried were broken (are nightly builds published even if the CI fails, or is the CI missing stuff to check that nightly builds are usable ?), so I give up.

/cc @jvican who might know what changed in zinc in 1.3 that could have caused this.

**Comment by eed3si9n** [comment]
Moved.

**Comment by jvican** [comment]
The root of the issue sounds scary. I would recommend disabling this kind of optimizations by default and not assume that sbt is using a concrete zinc version because that assumption is broken if sbt relies on bloop and receives analysis files from it (which is something it will soon do via our sbt-bloop plugin, which is used by Metals and is likely to increase in user adoption).

**Comment by smarter** [comment]
@eatkins That commit message says:

> In the future, I could see moving ExternalHooks into the zinc project so
that other tools like bloop or mill could use them.

I think this bug demonstrate that this would be a good idea indeed, why not do things in zinc to start with ?

**Comment by jvican** [comment]
> why not do things in zinc to start with ?
  
Honestly, I prefer this to be done in the Zinc users. Implementing it would increase the coupling with sbt-io, which I intend to remove by using NIO in all the Zinc APIs, and in-memory caching assumes a lot of things about how files are used by the build server, so it's better if this code lives in sbt/sbt.

**Comment by smarter** [comment]
Hmm surely it'd be good if zinc and sbt could agree on a strategy for file watching ?

**Comment by eatkins** [comment]
@jvican if we don't make it default, practically no one will use it and it will be hard to actually track down any bugs. I am, however, worried about breaking people's CI builds which is why the cache is only on for interactive sessions. I have been dogfooding the cache for almost a year and apart from occasionally seeing @smarter's issue, I have found it to work quite well 99% of the time.

Why should the zinc file stamp interface be subject to change? I should be able to implement xsbti.compile.analysis.Stamp and have zinc do the right thing with any stamps passed into it satisfying that interface. I don't think that is the case right now, but that seems like how it should be long term.

I 100% agree that we should decouple zinc and io. What I have been kicking around in my head is adding file reader and writer interfaces to the incremental options. The basic idea is we'd have something like
```java
interface xsbti.compile.analysis.io.FileReader {
  xsbti.compile.analysis.Stamp readStamp(File file)
  String readContent(File file, Charset charset)
}
```
```java
interface xsbti.compile.analysis.io.FileWriter {
  void writeStamp(File file)
  void writeFile(File file, Array[Byte] contents)
}
```
This would allow the application to choose how it wants to manage io. My guess is to implement something like this, I'd have to audit all of the io in zinc which is why I haven't looked at it yet.

**Comment by eatkins** [comment]
Also, the file system cache can be flushed with `clearCaches`. I also intend to make `clean` invoke `clearCaches` so that any cache inconsistencies can be resolved with a clean.

**Comment by smarter** [comment]
>  I am, however, worried about breaking people's CI builds which is why the cache is only on for interactive sessions

That sounds strange to me, is this temporary ? Having two code paths means twice as many bugs, and it also means that bugs in the CI might mysteriously not be reproducible when trying things interactively and vice-versa.

**Comment by smarter** [comment]
>  I have found it to work quite well 99% of the time.

As a member of the 1%, I can say this is definitely not good enough :).

**Comment by eatkins** [comment]
Yes. I understand that. What I am saying is that your bug does not mean that the cache is inherently error prone or will cause issues for most users. I have put a lot of thought into how to make everything easily configurable at runtime so that no one's build that previously worked will get unfixably broken by these changes. That is a big part of why these code reviews have been pending for months.

**Comment by jvican** [comment]
@eatkins Given that things are working for you guys right now I'd leave things as they are. I intend to rework the zinc APIs quite a bit during the next months and one of the first things I'd look at will be at using `Path` instead of `File` for any Zinc operation. When that change is done, it will most likely not be necessary to have `FileReader` or `FileWriter`s because, if you want that as a user, you should be able to write your own NIO File System.

**Comment by eatkins** [comment]
@jvican +1 for `Path` vs `File`. My concern is that if zinc actually calls say `Files.readAllBytes`, then there is no way to add a caching layer.

**Comment by jvican** [comment]
Indeed. We could just keep things as they are on that front so that you can implement the caching layer via `ExternalHooks`.

**Comment by eatkins** [comment]
@smarter I agree the default is a bit weird. I don't think it's so bad if I add a disclaimer on sbt startup for interactive builds that says something to the effect of "The sbt.io.FileTreeRepository implementation is unspecified, using a caching implementation. To remove this warning, set `Keys.fileTreeRepositoryImplementation := FileTreeRepositoryImplementation.Polling`." You and @jvican are absolutely right that silently using the cache with neither warning nor consent is a bug and should be fixed before the 1.3.x release.

**Comment by smarter** [comment]
I don't get it, either the caching is (believed to be) correct and no warning is needed, or the caching is known to be broken and it shouldn't be on by default. In any case the people working on sbt need to make that decision since they're the most well-informed as opposed to every single end-user having to make this decision themselves.

**Comment by eatkins** [comment]
Caching a file system subtree without polling is an extremely difficult problem to get right across all platforms. The data is inherently mutable and making sure that every incremental update is correctly applied and invokes the right callbacks for every update is not trivial at all. I have spent over a year working on it and while I am confident in my implementation, I cannot guarantee that there aren't edge cases that I haven't encountered. I can, however, guarantee that the cache reliably works for every use case that I have thought of and written tests for.

I think a disclaimer is a good idea until the feature stabilizes and has enough real world testing that we can be confident that it works for every important use case and the workarounds are well documented for scenarios in which it can't work. Like I said, I have done a ton of manual testing on mac/windows/linux (in that order), but I don't know what crazy nfs deployments, etc. are out there.

**Comment by eatkins** [comment]
Anyway, I'll take a look at @smarter's bug very soon and should be able to sort it out.

**Comment by smarter** [comment]
How about for 1.3.0 you run with caching enabled, but always check if the cached result matches the actual result: if it doesn't, print a big warning asking the user to report a bug. That way you can actually get data on how well it's working and eventually decide whether it can be used as the sole source of truth. Otherwise I'd be very wary of having this on by default, caching bugs can be incredibly hard to debug since they can cause very weird symptoms, I don't think a warning alone is going to be enough to mitigate that.

**Comment by eatkins** [comment]
That's a great idea.

**Comment by eatkins** [comment]
@smarter before I attempt to debug this issue, can you provide more system details? What version of linux are you using? What jdk?

**Comment by smarter** [comment]
See https://github.com/sbt/sbt/issues/4543#issuecomment-462966774

**Comment by eatkins** [comment]
Oh, whoops. Thanks

**Comment by eatkins** [comment]
I believe that this is fixed in 1.3.0-M2 and I will ensure that it stays fixed.

**Comment by fooblahblah** [comment]
I'm following this thread and trying to work around some similar issues on CircleCI where I get recompilations between stages of a parallel workflow. Anyway I'm wondering if the settings referred to above (https://github.com/sbt/sbt/issues/4543#issuecomment-462987046)  are still valid with sbt 1.3.0-RC1? I get errors with any of:
`Global/externalHooks := (_ => None)`
 or 
`externalHooks := (_ => None)`
or 
`Global/externalHooks := None`
or
`externalHooks := None`

```
sbt -batch test:compile
Picked up JAVA_TOOL_OPTIONS: -Dfile.encoding=UTF8 -Duser.timezone=UTC
[info] Loading global plugins from /home/jsimpson/.sbt/1.0/plugins
[info] Loading project definition from /home/jsimpson/workspace/sbt-inc-broken/project
/home/jsimpson/workspace/sbt-inc-broken/build.sbt:6: error: not found: value externalHooks
    Global/externalHooks := (_ => None)
```

**Comment by eatkins** [comment]
@fooblahblah the externalHooks setting was removed in https://github.com/sbt/sbt/commit/72df8f674c59cde5db7735ac1a2254c93d586bdd.
If you run with debug logging, you can see what files are triggering recompilations. It would be helpful to know if the unexpected recompilations are triggered by source files or binaries.

**Comment by fooblahblah** [comment]
Here's my fork from the project @smarter created with the changes I'm testing with: https://github.com/fooblahblah/sbt-inc-broken

I'm on Linux running OpenJDK 8:
```
$ java -version
Picked up JAVA_TOOL_OPTIONS: -Dfile.encoding=UTF8 -Duser.timezone=UTC
openjdk version "1.8.0_212"
OpenJDK Runtime Environment (build 1.8.0_212-ga)
OpenJDK 64-Bit Server VM (build 25.212-bga, mixed mode)
```
My sequence is as follows since I'm trying to simulate some issues I'm seeing in CircleCI to narrow down some sort of solution.
From directory the original directory I run:
```
$ sbt -batch test:compile
```
Then create a copy in `/tmp`  and build with the following commands:
```
$ cd .. && rm -rf sbt-inc-broken && cp -a ~/workspace/sbt-inc-broken . && cd sbt-inc-broken
$ sbt -batch test:compile
```
After that in my original directory a build is triggered which really seems confusing. Here's the debug output:
```
$ sbt -batch test:compile
Picked up JAVA_TOOL_OPTIONS: -Dfile.encoding=UTF8 -Duser.timezone=UTC
[info] Loading global plugins from /home/jsimpson/.sbt/1.0/plugins
[info] Loading project definition from /home/jsimpson/workspace/sbt-inc-broken/project
[info] Loading settings for project root from build.sbt ...
[info] Set current project to root (in build file:/home/jsimpson/workspace/sbt-inc-broken/)
[debug] Copy resource mappings: 
[debug] 	
[debug] 
[debug] Initial source changes: 
[debug] 	removed:Set()
[debug] 	added: Set()
[debug] 	modified: Set()
[debug] Invalidated products: Set(/home/jsimpson/workspace/sbt-inc-broken/target/scala-2.12/classes/A.class)
[debug] External API changes: API Changes: Set()
[debug] Modified binary dependencies: Set()
[debug] Initial directly invalidated classes: Set()
[debug] 
[debug] Sources indirectly invalidated by:
[debug] 	product: Set(/home/jsimpson/workspace/sbt-inc-broken/A.scala)
[debug] 	binary dep: Set()
[debug] 	external source: Set()
[debug] All sources are invalidated.
[debug] [inv] Invalidate package objects by inheritance only...
[debug] Initial set of included nodes: 
[debug] [inv] Package object invalidations: 
[debug] Recompiling all sources: number of invalidated sources > 50.0% of all sources
[debug] [inv] ********* Pruned: 
[debug] [inv] Relations (with name hashing enabled):
[debug] [inv]   products: Relation [ ]
[debug] [inv]   library deps: Relation [ ]
[debug] [inv]   library class names: Relation [
[debug] [inv]     /nix/store/13w0g5ysjmzxmwadjc6i37azzv1097hy-openjdk-8u212-ga/lib/openjdk/jre/lib/rt.jar -> java.lang.Object
[debug] [inv] ]
[debug] [inv]   internalDependencies: 
[debug] [inv]     
[debug] [inv]   externalDependencies: 
[debug] [inv]     
[debug] [inv]   class names: Relation [ ]
[debug] [inv]   used names: Relation [ ]
[debug] [inv]   product class names: Relation [ ]
[debug] [inv] *********
[info] Compiling 1 Scala source to /home/jsimpson/workspace/sbt-inc-broken/target/scala-2.12/classes ...
[debug] Getting org.scala-sbt:compiler-bridge_2.12:1.3.0-M5:compile for Scala 2.12.8
[debug] Getting org.scala-sbt:compiler-bridge_2.12:1.3.0-M5:compile for Scala 2.12.8
[debug] [zinc] Running cached compiler 6a04f61c for Scala compiler version 2.12.8
[debug] [zinc] The Scala compiler is invoked with:
[debug] 	-bootclasspath
[debug] 	/nix/store/13w0g5ysjmzxmwadjc6i37azzv1097hy-openjdk-8u212-ga/lib/openjdk/jre/lib/resources.jar:/nix/store/13w0g5ysjmzxmwadjc6i37azzv1097hy-openjdk-8u212-ga/lib/openjdk/jre/lib/rt.jar:/nix/store/13w0g5ysjmzxmwadjc6i37azzv1097hy-openjdk-8u212-ga/lib/openjdk/jre/lib/sunrsasign.jar:/nix/store/13w0g5ysjmzxmwadjc6i37azzv1097hy-openjdk-8u212-ga/lib/openjdk/jre/lib/jsse.jar:/nix/store/13w0g5ysjmzxmwadjc6i37azzv1097hy-openjdk-8u212-ga/lib/openjdk/jre/lib/jce.jar:/nix/store/13w0g5ysjmzxmwadjc6i37azzv1097hy-openjdk-8u212-ga/lib/openjdk/jre/lib/charsets.jar:/nix/store/13w0g5ysjmzxmwadjc6i37azzv1097hy-openjdk-8u212-ga/lib/openjdk/jre/lib/jfr.jar:/nix/store/13w0g5ysjmzxmwadjc6i37azzv1097hy-openjdk-8u212-ga/lib/openjdk/jre/classes:/home/jsimpson/.sbt/boot/scala-2.12.8/lib/scala-library.jar
[debug] 	-classpath
[debug] 	/home/jsimpson/workspace/sbt-inc-broken/target/scala-2.12/classes
[debug] Scala compilation took 2.386595531 s
[info] Done compiling.
[debug] [inv] ********* Fresh: 
[debug] [inv] Relations (with name hashing enabled):
[debug] [inv]   products: Relation [
[debug] [inv]     A.scala -> target/scala-2.12/classes/A.class
[debug] [inv] ]
[debug] [inv]   library deps: Relation [
[debug] [inv]     A.scala -> /nix/store/13w0g5ysjmzxmwadjc6i37azzv1097hy-openjdk-8u212-ga/lib/openjdk/jre/lib/rt.jar
[debug] [inv] ]
[debug] [inv]   library class names: Relation [
[debug] [inv]     /nix/store/13w0g5ysjmzxmwadjc6i37azzv1097hy-openjdk-8u212-ga/lib/openjdk/jre/lib/rt.jar -> java.lang.Object
[debug] [inv] ]
[debug] [inv]   internalDependencies: 
[debug] [inv]     
[debug] [inv]   externalDependencies: 
[debug] [inv]     
[debug] [inv]   class names: Relation [
[debug] [inv]     A.scala -> A
[debug] [inv] ]
[debug] [inv]   used names: Relation [
[debug] [inv]     A -> UsedName(A,[Default])
[debug] [inv]     A -> UsedName(AnyRef,[Default])
[debug] [inv]     A -> UsedName(Object,[Default])
[debug] [inv]     A -> UsedName(java;lang;Object;init;,[Default])
[debug] [inv]     A -> UsedName(scala,[Default])
[debug] [inv] ]
[debug] [inv]   product class names: Relation [
[debug] [inv]     A -> A
[debug] [inv] ]
[debug] [inv] *********
[debug] [inv] ********* Merged: 
[debug] [inv] Relations (with name hashing enabled):
[debug] [inv]   products: Relation [
[debug] [inv]     A.scala -> target/scala-2.12/classes/A.class
[debug] [inv] ]
[debug] [inv]   library deps: Relation [
[debug] [inv]     A.scala -> /nix/store/13w0g5ysjmzxmwadjc6i37azzv1097hy-openjdk-8u212-ga/lib/openjdk/jre/lib/rt.jar
[debug] [inv] ]
[debug] [inv]   library class names: Relation [
[debug] [inv]     /nix/store/13w0g5ysjmzxmwadjc6i37azzv1097hy-openjdk-8u212-ga/lib/openjdk/jre/lib/rt.jar -> java.lang.Object
[debug] [inv] ]
[debug] [inv]   internalDependencies: 
[debug] [inv]     
[debug] [inv]   externalDependencies: 
[debug] [inv]     
[debug] [inv]   class names: Relation [
[debug] [inv]     A.scala -> A
[debug] [inv] ]
[debug] [inv]   used names: Relation [
[debug] [inv]     A -> UsedName(A,[Default])
[debug] [inv]     A -> UsedName(AnyRef,[Default])
[debug] [inv]     A -> UsedName(Object,[Default])
[debug] [inv]     A -> UsedName(java;lang;Object;init;,[Default])
[debug] [inv]     A -> UsedName(scala,[Default])
[debug] [inv] ]
[debug] [inv]   product class names: Relation [
[debug] [inv]     A -> A
[debug] [inv] ]
[debug] [inv] *********
[debug] 
[debug] Initial source changes: 
[debug] 	removed:Set()
[debug] 	added: Set()
[debug] 	modified: Set()
[debug] Invalidated products: Set(/home/jsimpson/workspace/sbt-inc-broken/target/scala-2.12/test-classes/A.class)
[debug] External API changes: API Changes: Set()
[debug] Modified binary dependencies: Set()
[debug] Initial directly invalidated classes: Set()
[debug] 
[debug] Sources indirectly invalidated by:
[debug] 	product: Set(/home/jsimpson/workspace/sbt-inc-broken/A.scala)
[debug] 	binary dep: Set()
[debug] 	external source: Set()
[debug] All sources are invalidated.
[debug] [inv] Invalidate package objects by inheritance only...
[debug] Initial set of included nodes: 
[debug] [inv] Package object invalidations: 
[debug] Recompiling all sources: number of invalidated sources > 50.0% of all sources
[debug] [inv] ********* Pruned: 
[debug] [inv] Relations (with name hashing enabled):
[debug] [inv]   products: Relation [ ]
[debug] [inv]   library deps: Relation [ ]
[debug] [inv]   library class names: Relation [
[debug] [inv]     /nix/store/13w0g5ysjmzxmwadjc6i37azzv1097hy-openjdk-8u212-ga/lib/openjdk/jre/lib/rt.jar -> java.lang.Object
[debug] [inv] ]
[debug] [inv]   internalDependencies: 
[debug] [inv]     
[debug] [inv]   externalDependencies: 
[debug] [inv]     
[debug] [inv]   class names: Relation [ ]
[debug] [inv]   used names: Relation [ ]
[debug] [inv]   product class names: Relation [ ]
[debug] [inv] *********
[info] Compiling 1 Scala source to /home/jsimpson/workspace/sbt-inc-broken/target/scala-2.12/test-classes ...
[debug] Getting org.scala-sbt:compiler-bridge_2.12:1.3.0-M5:compile for Scala 2.12.8
[debug] Getting org.scala-sbt:compiler-bridge_2.12:1.3.0-M5:compile for Scala 2.12.8
[debug] [zinc] Running cached compiler 7addf128 for Scala compiler version 2.12.8
[debug] [zinc] The Scala compiler is invoked with:
[debug] 	-bootclasspath
[debug] 	/nix/store/13w0g5ysjmzxmwadjc6i37azzv1097hy-openjdk-8u212-ga/lib/openjdk/jre/lib/resources.jar:/nix/store/13w0g5ysjmzxmwadjc6i37azzv1097hy-openjdk-8u212-ga/lib/openjdk/jre/lib/rt.jar:/nix/store/13w0g5ysjmzxmwadjc6i37azzv1097hy-openjdk-8u212-ga/lib/openjdk/jre/lib/sunrsasign.jar:/nix/store/13w0g5ysjmzxmwadjc6i37azzv1097hy-openjdk-8u212-ga/lib/openjdk/jre/lib/jsse.jar:/nix/store/13w0g5ysjmzxmwadjc6i37azzv1097hy-openjdk-8u212-ga/lib/openjdk/jre/lib/jce.jar:/nix/store/13w0g5ysjmzxmwadjc6i37azzv1097hy-openjdk-8u212-ga/lib/openjdk/jre/lib/charsets.jar:/nix/store/13w0g5ysjmzxmwadjc6i37azzv1097hy-openjdk-8u212-ga/lib/openjdk/jre/lib/jfr.jar:/nix/store/13w0g5ysjmzxmwadjc6i37azzv1097hy-openjdk-8u212-ga/lib/openjdk/jre/classes:/home/jsimpson/.sbt/boot/scala-2.12.8/lib/scala-library.jar
[debug] 	-classpath
[debug] 	/home/jsimpson/workspace/sbt-inc-broken/target/scala-2.12/test-classes:/home/jsimpson/workspace/sbt-inc-broken/target/scala-2.12/classes
[debug] Scala compilation took 0.373693734 s
[info] Done compiling.
[debug] [inv] ********* Fresh: 
[debug] [inv] Relations (with name hashing enabled):
[debug] [inv]   products: Relation [
[debug] [inv]     A.scala -> target/scala-2.12/test-classes/A.class
[debug] [inv] ]
[debug] [inv]   library deps: Relation [
[debug] [inv]     A.scala -> /nix/store/13w0g5ysjmzxmwadjc6i37azzv1097hy-openjdk-8u212-ga/lib/openjdk/jre/lib/rt.jar
[debug] [inv] ]
[debug] [inv]   library class names: Relation [
[debug] [inv]     /nix/store/13w0g5ysjmzxmwadjc6i37azzv1097hy-openjdk-8u212-ga/lib/openjdk/jre/lib/rt.jar -> java.lang.Object
[debug] [inv] ]
[debug] [inv]   internalDependencies: 
[debug] [inv]     
[debug] [inv]   externalDependencies: 
[debug] [inv]     
[debug] [inv]   class names: Relation [
[debug] [inv]     A.scala -> A
[debug] [inv] ]
[debug] [inv]   used names: Relation [
[debug] [inv]     A -> UsedName(A,[Default])
[debug] [inv]     A -> UsedName(AnyRef,[Default])
[debug] [inv]     A -> UsedName(Object,[Default])
[debug] [inv]     A -> UsedName(java;lang;Object;init;,[Default])
[debug] [inv]     A -> UsedName(scala,[Default])
[debug] [inv] ]
[debug] [inv]   product class names: Relation [
[debug] [inv]     A -> A
[debug] [inv] ]
[debug] [inv] *********
[debug] [inv] ********* Merged: 
[debug] [inv] Relations (with name hashing enabled):
[debug] [inv]   products: Relation [
[debug] [inv]     A.scala -> target/scala-2.12/test-classes/A.class
[debug] [inv] ]
[debug] [inv]   library deps: Relation [
[debug] [inv]     A.scala -> /nix/store/13w0g5ysjmzxmwadjc6i37azzv1097hy-openjdk-8u212-ga/lib/openjdk/jre/lib/rt.jar
[debug] [inv] ]
[debug] [inv]   library class names: Relation [
[debug] [inv]     /nix/store/13w0g5ysjmzxmwadjc6i37azzv1097hy-openjdk-8u212-ga/lib/openjdk/jre/lib/rt.jar -> java.lang.Object
[debug] [inv] ]
[debug] [inv]   internalDependencies: 
[debug] [inv]     
[debug] [inv]   externalDependencies: 
[debug] [inv]     
[debug] [inv]   class names: Relation [
[debug] [inv]     A.scala -> A
[debug] [inv] ]
[debug] [inv]   used names: Relation [
[debug] [inv]     A -> UsedName(A,[Default])
[debug] [inv]     A -> UsedName(AnyRef,[Default])
[debug] [inv]     A -> UsedName(Object,[Default])
[debug] [inv]     A -> UsedName(java;lang;Object;init;,[Default])
[debug] [inv]     A -> UsedName(scala,[Default])
[debug] [inv] ]
[debug] [inv]   product class names: Relation [
[debug] [inv]     A -> A
[debug] [inv] ]
[debug] [inv] *********
[debug] Copy resource mappings: 
[debug] 	
[debug] Framework implementation 'org.scalacheck.ScalaCheckFramework' not present.
[debug] Framework implementation 'org.specs2.runner.Specs2Framework' not present.
[debug] Framework implementation 'org.specs2.runner.SpecsFramework' not present.
[debug] Framework implementation 'org.specs.runner.SpecsFramework' not present.
[debug] Framework implementation 'org.scalatest.tools.Framework' not present.
[debug] Framework implementation 'org.scalatest.tools.ScalaTestFramework' not present.
[debug] Framework implementation 'com.novocode.junit.JUnitFramework' not present.
[debug] Subclass fingerprints: List()
[debug] Annotation fingerprints: List()
[success] Total time: 4 s, completed May 28, 2019 4:45:21 PM
```

**Comment by eatkins** [comment]
This looks like a duplicate of https://github.com/sbt/sbt/issues/4704. Thanks for providing the sample project. It will be helpful for me to track down the issue.

**Comment by fooblahblah** [comment]
Cool, I don't know if I've seen that issue. I'll keep tabs over there.

**Comment by fooblahblah** [comment]
At a more basic level it's affecting me in 1.2.8 and we're pretty much blocked trying to parallel-ize our CircleCI build which occurs on various containers and I'm trying to coordinate by copying around `target` or whatever I need to prevent compilation of "already compiled" resources. Any other issues I should check out?

**Comment by eatkins** [comment]
That sounds like a hard use case for sbt to support. There is an implicit assumption in the way that sbt uses zinc that all the product files for the build are managed by sbt. If there is out of band file copying, I'd expect there to be issues with incremental compilation. With the caveat that I don't know what you're doing, my suggestion is that suppose that you are copying around target directories, you should add an sbt task that does the copying, say, `val copyClassFiles = taskKey[Unit]("...")`, then you could add `compile in Compile := (compile in Compile).dependsOn(copyClassFiles).value.` That would ensure that the copying is fully done before sbt tries to get the file stamps.

**Comment by eatkins** [comment]
sbt 1.2.8 does not use custom external hooks so any problems you are having there would be unrelated to #4704.

**Comment by fooblahblah** [comment]
Hmm, interesting. I'll try and pursue that file copying avenue for a bit before giving up.

I was originally using 1.2.8 since that's the version called for from Play-2.7.2 (which we're also using). I was only trying to use 1.3.0-RC1 in hopes that my original issue of moving `target` between builds would work.

**Comment by jvican** [comment]
@fooblahblah You could support that use case with [Bloop](https://github.com/scalacenter/bloop) (via the sbt plugin sbt-bloop). I know several people using similar tricks in the CI to speed up their builds. Feel free to drop by https://gitter.im/scalacenter/bloop if you have any questions about how to approach this problem

**Comment by fooblahblah** [comment]
@jvican Very good to know! I'll check that out.

**Comment by eatkins** [comment]
@fooblahblah I was thinking a bit more about your issue and I think copying around target directories may be a problematic strategy for parallelization. The reason is that zinc checks whether the class files that your project depends on have changed since the last run. This check uses the last modified time of each class file (and dependency jar). If you are copying the class files every time, that will change the last modified of all of the files which effectively invalidates the entire class path.

**Comment by smarter** [comment]
You can copy files without changing the last modified date (e.g. by using `cp -a`)

**Comment by eatkins** [comment]
Aah, cool. Didn't think of that.

**Comment by fooblahblah** [comment]
I was going to follow-up when I had completed this CI task and it was actually working, but I'm having success tar'ing between build steps using a little `posix` flag which keep timestamps down to the nanosecond (I think). 

```
tar -H posix -czvf ../project.tgz .
```
I was trying to only archive `target` to save a little time, but from some tinkering I've found best results where I use the above command after the build phase. Then downstream I un-tar the whole tree and SBT seems to chug along with the test without compiling.

The command for the tests is:
```
./sbt -J-Xmx14g "set offline := true" "set skip in update := true" test
```

I'm still working out a few details, but it's looking promising. This is with sbt-1.2.8.

Just for background I'm trying to use CircleCI workflows so I can parallel-ize our webpack and scala builds, then run the tests, then deploy to Elastic Beanstalk and ECR in parallel too. Without the archiving I'd get hit by a recompile each time.

Bloop looks promising, but we're currently dependent on a couple SBT plugins I'd need to extract into scripts before bloop will be usable for us.
